common/shell.hpp

```diff
@@ -30,58 +30,63 @@ inline process::Future<std::string> runCommand(
 
   if (s.isError()) {
     return process::Failure(
-      "Unable to execute '" + command + "': " + s.error());
+        "Unable to execute '" + command + "': " + s.error());
   }
   pid_t pid = s->pid();
 
   typedef std::tuple<
-    process::Future<Option<int>>,
-    process::Future<std::string>,
-    process::Future<std::string>> ProcessTuple;
+      process::Future<Option<int>>,
+      process::Future<std::string>,
+      process::Future<std::string>> ProcessTuple;
 
   return await(
       s->status(),
       process::io::read(s->out().get()),
       process::io::read(s->err().get()))
-    .after(timeout,
-                [=](const process::Future<ProcessTuple> &t) ->
-                    process::Future<ProcessTuple> {
-        // NOTE: Discarding this future has no effect on the subprocess
-        os::killtree(pid, SIGKILL);
-        return std::make_tuple(-1, "", "timeout after " + stringify(timeout));
+    .after(timeout, [=](const process::Future<ProcessTuple> &future) ->
+        process::Future<ProcessTuple> {
+      // NOTE: Discarding this future has no effect on the subprocess
+      os::killtree(pid, SIGKILL);
+      return future.then(
```

Comment:
The process has gone and the reaper is done signaled by those inner futures (`ProcessTuple`) getting into a terminal state. In turn `await`s future gets ready, so we hit `then` - that is when we can finally exit.